runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -166,11 +168,20 @@ public static DefaultJobBundleFactory create(
         CacheBuilder.newBuilder()
             .removalListener(
                 (RemovalNotification<Environment, WrappedSdkHarnessClient> notification) -> {
-                  int refCount = notification.getValue().unref();
-                  LOG.debug(
-                      "Removed environment {} with {} remaining bundle references.",
-                      notification.getKey(),
-                      refCount);
+                  WrappedSdkHarnessClient client = notification.getValue();
```

Comment:
Having this as part of the removal listener would prematurely close an environment that is still referenced. The purpose of the refcount is to be able to remove the environment from the cache when it expires but close it only after all references are gone.

Reply:
It doesn't work though if we do not ensure dereferencing under all circumstances. We need a safeguard here, also considering other runners may not dereference correctly. Generally, it is hard to guarantee dereferencing due to the nesting of DoFnRunners which may not even allow closing the bundle in error cases. I considered not doing this but I think it is the safer route.

If you take a step back, when would the reference counting really be useful? Every restarted job will anyways run in a new classloader, so the environment will never be recycled. When we call close we should tear down everything. 

Taking back another step, the reference counting should really be removed entirely. It was error prone from the beginning leading to subtle problems with dereferencing. If you don't mind, I'd remove it. What do you think?

Reply:
I've removed the reference counting, PTAL.

Reply:
As stated above, reference counting is required for environment expiration. The environment can only be closed when all bundles that reference it have finished. I would prefer we limit this PR scope strictly to the original purpose and discuss other changes separately.

Reply:
The correct solution might be to enforce this from close, not from the cache removal listener.

Reply:
You are right that we can't close the environment during removal due to pending references hold by still-processing bundles. Please take a look at the follow-up. I'm now quarantining clients which still hold references, to remove them during shutdown if they have not been dereferenced before.